File: z4.cpp
#include <iostream>
#include <string>
#include <map>
#include <stdexcept>
#include <limits>

using namespace std;

map<string, string> regions;

void change(const string& region, const string& center) {
    if (regions.find(region) != regions.end()) {
        // Переименование центра
        if (regions[region] == center) {
            throw runtime_error("Невозможно переименовать " + center + " в " + center);
        } else {
            string old_center = regions[region];
            regions[region] = center;
            cout << "Region " << region << " has changed its administrative center from "
                 << old_center << " to " << center << endl;
        }
    } else {
        // Создание нового региона
        regions[region] = center;
        cout << "New region " << region << " with administrative center " << center << endl;
    }
}

void rename(const string& old_region, const string& new_region) {
    if (regions.find(old_region) == regions.end()) {
        throw runtime_error("Региона с названием " + old_region + " не существует!");
    } else if (regions.find(new_region) != regions.end()) {
        throw runtime_error("Регион с названием " + new_region + " уже существует!");
    } else {
        string center = regions[old_region];
        regions.erase(old_region);
        regions[new_region] = center;
        cout << old_region << " has been renamed to " << new_region << endl;
    }
}

void about(const string& region) {
    if (regions.find(region) != regions.end()) {
        cout << region << " has administrative center " << regions[region] << endl;
    } else {
        throw runtime_error("Региона с названием " + region + " не существует!");
    }
}

void all() {
    if (!regions.empty()) {
        cout << "Список регионов:" << endl;
        for (map<string, string>::const_iterator it = regions.begin(); it != regions.end(); ++it) {
            cout << it->first << " - " << it->second << endl;
        }
    } else {
        throw runtime_error("Список регионов пуст!");
    }
}

int main() {
    int N;
    cout << "Введите количество запросов > ";
    cin >> N;

    for (int i = 0; i < N; ++i) {
        string cmd;
        cout << "   Введите команду > ";
        cin >> cmd;

        try {
            if (cmd == "CHANGE") {
                string region, center;
                cin >> region >> center;
                change(region, center);
            } else if (cmd == "RENAME") {
                string old_region, new_region;
                cin >> old_region >> new_region;
                rename(old_region, new_region);
            } else if (cmd == "ABOUT") {
                string region;
                cin >> region;
                about(region);
            } else if (cmd == "ALL") {
                all();
            } else {
                cin.clear();
                cin.ignore(numeric_limits<streamsize>::max(), '\n');
                cout << "Введена неверная команда!" << endl;
            }
        } catch (const exception& e) {
            cerr << e.what() << endl;
        }
    }

   return 0;
}

File: z4.py
regions = {}

def change(region, center):
    if(region in regions):
        #переименование центра
        if (center == regions[region]):
            print(f'Невозможно переименовать {center} в {center}')
        else:
            old_center = regions[region]
            regions[region] = center
            print(f'Region {region} has changed its administrative center from {old_center} to {center}')
    else:
        #создание региона
        regions[region] = center
        print(f'New region {region} with administrative center {center}')

def rename(old_region, new_region):
    if (old_region in regions):
        if (new_region in regions):
            print(f'Регион с названием {new_region} уже существует!')
        else:
            center = regions[old_region]
            regions.pop(old_region)
            regions[new_region] = center
            print(f'{old_region} has been renamed to {new_region}')
    else:
        print(f'Региона с названием {old_region} не существует!')

def about(region):
    if (region in regions):
        print(f'{region} has administrative center {regions[region]}')
    else:
        print(f'Региона с названием {region} не существует!')

def all():
    if len(regions) > 0:
        print("Список регионов:")
        for region, center in regions.items():
            print(f'{region} - {center}')
    else:
        print("Список регионов пуст!")


N = int(input("Введите количество запросов > "))
for i in range(N):
    cmd = input("   Введите команду > ").split()
    if cmd[0] == "CHANGE" and len(cmd) == 3:
        change(cmd[1], cmd[2])
    elif cmd[0] == "RENAME" and len(cmd) == 3:
        rename(cmd[1], cmd[2])
    elif cmd[0] == "ABOUT" and len(cmd) == 2:
        about(cmd[1])
    elif cmd[0] == "ALL" and len(cmd) == 1:
        all()
    else:
        print("Введена неверная команда!")